server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,296 @@
+            */
+            fileCache.remove(localDirectory.getDirectory().resolve(name));
+        }
+    }
+
+    /**
+     * Returns the byte length of a file in the directory.
+     * Throws {@link NoSuchFileException} or {@link FileNotFoundException} in case file is not present locally and in remote as well
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public long fileLength(String name) throws IOException {
+        logger.trace("fileLength() called {}", name);
+        long fileLength;
+        Path key = localDirectory.getDirectory().resolve(name);
+        if (isTempFile(name) || fileCache.get(key) != null) {
+            try {
+                fileLength = localDirectory.fileLength(name);
+                logger.trace("fileLength from Local {}", fileLength);
+            } finally {
+                fileCache.decRef(key);
+            }
+        } else {
+            fileLength = remoteDirectory.fileLength(name);
+            logger.trace("fileLength from Remote {}", fileLength);
+        }
+        return fileLength;
+    }
+
+    /**
+     * Creates a new, empty file in the directory and returns an {@link IndexOutput} instance for
+     * appending data to this file.
+     * @param name the name of the file to create.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public IndexOutput createOutput(String name, IOContext context) throws IOException {
+        logger.trace("createOutput() called {}", name);
+        // The CloseableFilterIndexOutput will ensure that the file is added to FileCache once write is completed on this file
+        return new CloseableFilterIndexOutput(localDirectory.createOutput(name, context), name, this::cacheFile);
+    }
+
+    /**
+     * Ensures that any writes to these files are moved to stable storage (made durable).
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void sync(Collection<String> names) throws IOException {
+        logger.trace("sync() called {}", names);
+        Collection<String> remoteFiles = Arrays.asList(getRemoteFiles());
+        Collection<String> filesToSync = names.stream().filter(name -> remoteFiles.contains(name) == false).collect(Collectors.toList());
+        logger.trace("Synced files : {}", filesToSync);
+        localDirectory.sync(filesToSync);
+    }
+
+    /**
+     * Renames {@code source} file to {@code dest} file where {@code dest} must not already exist in
+     * the directory.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void rename(String source, String dest) throws IOException {
+        logger.trace("rename() called {}, {}", source, dest);
+        localDirectory.rename(source, dest);
+        fileCache.remove(localDirectory.getDirectory().resolve(source));
+        cacheFile(dest);
```

Comment:
I think you will need to `decRef` for `dest` here since `cacheFile` will `incRef` on that file.

Reply:
Don't see the change for `decRef`, is it not needed ?

Reply:
decRef will be done after the newly renamed file is uploaded to remote